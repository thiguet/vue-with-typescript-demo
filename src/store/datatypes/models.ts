interface Unique {
    id?: string;
}

export enum Measures {
    UN = 'UN',
    L = 'L',
    KG = 'KG',
    G = 'G',
}

export interface Product extends Unique {
    name: string;
    measure: Measures;
    qtd: number;
    minQtd: number;
    image?: string;
}

export interface User extends Unique {
    name: string;
    email: string;
}

export interface ReportItem {
    text: string;
    value: number;
}

export type ReportData = ReportItem[];

export interface QuantityReportData {
    productName: string;
    productQuantity: number;
}

export type QuantityReport = QuantityReportData[];
export interface MeasuresReportData {
    measureName: string;
    measureQuantity: number;
}

export type MeasuresReport = MeasuresReportData[];

export enum VuexAppModules {
    products = 'products',
    login = 'login',
    alert = 'alert',
    reports = 'reports',
}
