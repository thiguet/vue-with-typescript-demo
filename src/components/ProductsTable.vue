<template>
    <div class="table-container">
        <table id="table" width="100%">
            <tbody>
                <tr
                    class="row"
                    v-for="({ icon, name }, index) in rows"
                    :key="index"
                >
                    <td class="left">
                        <span><img :src="icon"/></span>
                    </td>
                    <td>
                        <span>{{ name }}</span>
                    </td>
                    <td class="right">
                        <div>
                            <Button
                                :id="`edit-${index + 1}`"
                                class="edit"
                                icon="./assets/icons/edit.svg"
                                name="editar"
                                label="Edit"
                                :onclick="evt => onEdit(evt, index)"
                            />
                            <Button
                                :id="`delete-${index + 1}`"
                                class="delete"
                                icon="./assets/icons/delete.svg"
                                name="deletar"
                                label="Delete"
                                :onclick="evt => onDelete(evt, index)"
                            />
                        </div>
                    </td>
                </tr>
            </tbody>
        </table>
    </div>
</template>

<script lang="ts">
import Vue from 'vue';

import { ProductsTableComp } from '@/components/models.d';

import Component from 'vue-class-component';

import Button from '@/components/Button.vue';
import { Prop } from 'vue-property-decorator';

interface RowsData {
    icon: string;
    name: string;
}

@Component({
    name: 'ProductsTable',
    components: {
        Button,
    },
})
export default class ProductsTable extends Vue implements ProductsTableComp {
    @Prop({ required: true }) private rows!: Array<RowsData>;

    public onEdit(evt: Event, index: number) {
        return this.$emit('on-edit', index);
    }

    public onDelete(evt: Event, index: number) {
        return this.$emit('on-delete', index);
    }
}
</script>

<style>
#table {
    margin: auto;
    width: 100%;
}

#table > tbody > tr > td > span > img {
    width: 60px;
    filter: brightness(0);
}

.left {
    text-align: 'left';
    margin-left: 1em;
}
.right {
    text-align: 'right';
    margin-right: 1em;
}

.edit {
    justify-content: flex-end !important;
}

.edit > button,
.delete > button {
    background: none;
}

.icon {
    width: 40px;
}
</style>
