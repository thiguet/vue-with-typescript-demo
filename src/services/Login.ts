import { LoginSubmit, User } from '../store/datatypes/models';

import VueWithTSAPI from './index';

export { VueWithTSAPI };

export const setJWT = (jwt: string) => {
    VueWithTSAPI.defaults.headers.common.Authorization = `Bearer ${jwt}`;
};

export const clearJWT = () => {
    VueWithTSAPI.defaults.headers.common.Authorization = null;
};

export async function login(user: LoginSubmit): Promise<User> {
    const response = await VueWithTSAPI.post('/users/login', {
        user,
    });
    return response.data as User;
}
