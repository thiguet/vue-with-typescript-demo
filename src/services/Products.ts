import { Product } from '@/store/datatypes/models';

import VueWithTSAPI from './index';

export { VueWithTSAPI };

export async function getAllProducts(): Promise<Product[]> {
    return (await VueWithTSAPI.get('/products')).data as Product[];
}

export async function newProduct(product: Product): Promise<Product> {
    return (
        await VueWithTSAPI.post('/products', {
            product,
        })
    ).data as Product;
}

export async function editProduct(product: Product): Promise<Product> {
    return (
        await VueWithTSAPI.put(`/products/${product.id}`, {
            product,
        })
    ).data as Product;
}

export async function deleteProduct(product: Product): Promise<Product> {
    return (await VueWithTSAPI.delete(`/products/${product.id}`))
        .data as Product;
}
