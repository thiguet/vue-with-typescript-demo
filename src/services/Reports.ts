import { QuantityReport, MeasuresReport } from '@/store/datatypes/models';

import VueWithTSAPI from '@/services/';

export { VueWithTSAPI };

export async function getQuantityReport(): Promise<QuantityReport> {
    return (await VueWithTSAPI.get('/reports/quantity')).data as QuantityReport;
}

export async function getMeasuresReport(): Promise<MeasuresReport> {
    return (await VueWithTSAPI.get('/reports/measures')).data as MeasuresReport;
}
