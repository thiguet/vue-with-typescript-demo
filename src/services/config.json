{
    "baseURL": "http://localhost:8090"
}
