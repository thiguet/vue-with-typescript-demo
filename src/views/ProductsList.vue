<template>
    <div class="products-container">
        <div class="row">
            <ProductsTable
                :rows="rows"
                @on-edit="routeToEditProductPage"
                @on-delete="removeLine"
            />
        </div>
        <div class="row">
            <Button
                id="go-back"
                name="voltar"
                label="Voltar"
                icon="./assets/icons/goback.svg"
                class="footer-btn"
                :onclick="routeToHomePage"
            />
            <Button
                id="add-product"
                name="Add Product"
                label="Add Product"
                icon="./assets/icons/add.svg"
                class="footer-btn"
                :onclick="routeToNewProductPage"
            />
        </div>
    </div>
</template>

<script lang="ts">
import { Vue, Component } from 'vue-property-decorator';

import Button from '@/components/Button.vue';

import ProductsTable from '@/components/ProductsTable.vue';

import { ProductsListView } from './models.d';

interface RowsData {
    name: string;
    icon: string;
}

@Component({
    name: 'ProductsList',
    components: {
        Button,
        ProductsTable,
    },
    data() {
        return {
            rows: [
                {
                    name: 'test 1',
                    icon: './assets/icons/statistics.svg',
                },
                {
                    name: 'test 2',
                    icon: './assets/icons/statistics.svg',
                },
                {
                    name: 'test 3',
                    icon: './assets/icons/statistics.svg',
                },
                {
                    name: 'test 4',
                    icon: './assets/icons/statistics.svg',
                },
                {
                    name: 'test 5',
                    icon: './assets/icons/statistics.svg',
                },
            ],
        };
    },
})
export default class ProductsList extends Vue implements ProductsListView {
    private rows!: Array<RowsData>;

    public routeToHomePage() {
        this.$router.push('/');
    }

    public routeToNewProductPage() {
        this.$router.push('/products/new');
    }

    public routeToEditProductPage() {
        this.$router.push('/products/edit');
    }

    public removeLine(index: number) {
        this.rows.splice(index, 1);
    }
}
</script>

<style>
.products-container {
    flex-direction: column;
}

.row:nth-child(2) {
    margin-top: 40px;
    margin-bottom: 40px;
}

.footer-btn > button {
    padding: 5px 50px;
}
.footer-btn > button > img {
    display: flex;
    width: 50px;
}
</style>
