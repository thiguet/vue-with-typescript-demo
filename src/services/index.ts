import axios from 'axios';

import config from './config.json';

export default axios.create({
    baseURL: config.baseURL,
    headers: {
        Accept: 'application/json',
    },
});
