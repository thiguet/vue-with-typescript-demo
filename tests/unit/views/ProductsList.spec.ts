import { shallowMount, mount, createLocalVue, Wrapper } from '@vue/test-utils';

import Vuex, { ModuleTree, Store } from 'vuex';
import VueRouter from 'vue-router';

import ProductsList from '@/views/ProductsList.vue';
import ProductsTable from '@/components/ProductsTable.vue';
import faker from 'faker';
import { VuexAppModules } from '@/store/datatypes/models';
import {
    State,
    MutationTypes,
    ActionTypes,
    GettersTypes,
} from '@/store/modules/products';
import { getAllProducts } from '@/services/Products';
import { ProductsVuex } from '../store/models.d';
import { getFakeProduct, getFakeProductArray } from '../utils/ProductFactory';

jest.mock('@/services/Products');

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(VueRouter);

const router = new VueRouter();

describe('ProductsList', () => {
    let store: Store<State>;

    let products: ProductsVuex;

    const build = () => {
        const options = {
            localVue,
            store,
            router,
        };

        const wrapper: Wrapper<ProductsList> = shallowMount(
            ProductsList,
            options,
        );
        const mountedWrapper: Wrapper<ProductsList> = mount(
            ProductsList,
            options,
        );

        return {
            wrapper,
            productsList: () => mountedWrapper.findComponent(ProductsTable),
            goBackButton: () => mountedWrapper.find('#go-back'),
            addProductButton: () => mountedWrapper.find('#add-product'),
        };
    };

    beforeEach(() => {
        router.push = jest.fn();

        products = {
            namespaced: true,
            state: {
                editMode: false,
                selectedProduct: {
                    ...getFakeProduct(),
                },
                products: getFakeProductArray(),
            },
            getters: {
                [GettersTypes.tableRows]: jest.fn(),
            },
            mutations: {
                [MutationTypes.setProductName]: jest.fn(),
                [MutationTypes.setProductQtd]: jest.fn(),
                [MutationTypes.setProductMinQtd]: jest.fn(),
                [MutationTypes.setProductMeasure]: jest.fn(),
                [MutationTypes.setProductImage]: jest.fn(),
            },
            actions: {
                [ActionTypes.newProduct]: jest.fn(),
                [ActionTypes.editProduct]: jest.fn(),
                [ActionTypes.saveProduct]: jest.fn(),
                [ActionTypes.fetchProducts]: jest.fn(),
                [ActionTypes.selectProduct]: jest.fn(),
                [ActionTypes.deleteProduct]: jest.fn(),
            },
        };

        const modules: ModuleTree<State> = {
            [VuexAppModules.products]: {
                ...products,
            },
        };

        store = new Vuex.Store({
            modules,
        });
    });

    it('renders component', () => {
        const { wrapper } = build();
        expect(wrapper).toMatchSnapshot();
    });

    it('renders main components', () => {
        const { wrapper } = build();
        expect(wrapper).toMatchSnapshot();
    });

    it('fetch products on mount', async () => {
        build();

        await setTimeout(() => {
            expect(getAllProducts).toHaveBeenCalled();
        }, 1000);
    });

    it('go back to home page', async () => {
        const { goBackButton } = build();

        await goBackButton().trigger('click');

        expect(router.push).toHaveBeenCalledWith('/');
    });

    it('go back to new product page', async () => {
        const { actions } = products;

        const { addProductButton } = build();

        await addProductButton().trigger('click');

        expect(actions.newProduct).toBeCalled();

        expect(router.push).toHaveBeenCalledWith('/products/new');
    });

    it('sends the user to the edit page selecting the product', async () => {
        const { actions } = products;

        const { productsList } = build();

        const index = faker.random.number({ min: 0, max: 5 });

        await productsList().vm.$emit('on-edit', index);

        expect(router.push).toHaveBeenCalledWith('/products/edit');

        expect(actions.editProduct).toBeCalled();

        expect(actions.selectProduct).toBeCalled();
    });

    it('calls deleteProduct action', async () => {
        const { actions } = products;

        const { productsList } = build();

        const index = faker.random.number({ min: 0, max: 5 });

        await productsList().vm.$emit('on-delete', index);

        expect(actions.deleteProduct).toBeCalled();
    });
});
