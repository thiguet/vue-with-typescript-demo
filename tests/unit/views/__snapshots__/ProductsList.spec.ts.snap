// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsList renders component 1`] = `
<div class="products-container">
  <div class="row">
    <productstable-stub></productstable-stub>
  </div>
  <div class="row">
    <button-stub id="go-back" name="voltar" onclick="function () { [native code] }" label="Voltar" icon="./assets/icons/goback.svg" class="footer-btn"></button-stub>
    <button-stub id="add-product" name="Add Product" onclick="function () { [native code] }" label="Add Product" icon="./assets/icons/add.svg" class="footer-btn"></button-stub>
  </div>
</div>
`;

exports[`ProductsList renders main components 1`] = `
<div class="products-container">
  <div class="row">
    <productstable-stub></productstable-stub>
  </div>
  <div class="row">
    <button-stub id="go-back" name="voltar" onclick="function () { [native code] }" label="Voltar" icon="./assets/icons/goback.svg" class="footer-btn"></button-stub>
    <button-stub id="add-product" name="Add Product" onclick="function () { [native code] }" label="Add Product" icon="./assets/icons/add.svg" class="footer-btn"></button-stub>
  </div>
</div>
`;
